File: src/com/leozhi/topic401/Solution03.java
package com.leozhi.topic401;

import java.util.ArrayList;
import java.util.List;

/**
 * @author leozhi
 * 回溯
 * 通过
 * 7ms
 */
public class Solution03 {
    List<String> res;
    public List<String> readBinaryWatch(int num) {
        res = new ArrayList<>();
        if (num == 0) {
            res.add("0:00");
            return res;
        }
        String trace = "";
        backtrace(trace, num);
        return res;
    }

    private void backtrace(String trace, int num) {
        if (trace.length() > 10) {
            return;
        }
        if (num == 0) {
            int hour, minute;
            if (trace.length() <= 4) {
                hour = Integer.parseInt(trace, 2) << 4 - trace.length();
                minute = 0;
            } else {
                hour = Integer.parseInt(trace.substring(0, 4), 2);
                minute = Integer.parseInt(trace.substring(4), 2) << 10 - trace.length();
            }
            if (hour > 11 || minute > 59) {
                return;
            }
            String time;
            if (minute < 10) {
                time = hour + ":0" + minute;
            } else {
                time = hour + ":" + minute;
            }
            res.add(time);
            return;
        }
        // 选择 1
        trace += "1";
        backtrace(trace, --num);
        trace = trace.substring(0, trace.length() - 1);
        ++num;
        // 选择 0
        trace += "0";
        backtrace(trace, num);
    }
}

File: src/com/leozhi/topic13/Solution03.kt
package com.leozhi.topic13

/**
 * @author leozhi
 * 哈希表
 * 通过
 * 240ms
 */
class Solution03 {
    fun romanToInt(s: String): Int {
        val len = s.length
        val map = hashMapOf('I' to 1, 'V' to 5, 'X' to 10, 'L' to 50, 'C' to 100, 'D' to 500, 'M' to 1000)
        var res = 0
        for (i in 0 until len) {
            res += map[s[i]] ?: 0
            if (i > 0) {
                when {
                    s[i - 1] == 'I' && (s[i] == 'V' || s[i] == 'X') -> res -= 2
                    s[i - 1] == 'X' && (s[i] == 'L' || s[i] == 'C') -> res -= 20
                    s[i - 1] == 'C' && (s[i] == 'D' || s[i] == 'M') -> res -= 200
                }
            }
        }
        return res
    }
}

fun main() {
    for (i in 1..3999) {
        println(i)
    }
}